Devel nginx https

matt9j opened the pull request:
This pull request address #23 and has been tested in the uw deployment of CCM. Please advise us on commit structure, upstreaming procedures, etc. if there's anything else we should add/change!
Cheers,
-mj

Commit:
[cloud] Reject non-https nginx posts

Summary: Nginx would attempt to redirect non-https posts, which is
highly discouraged and leads to odd behavior.

Commit:
[cloud] Fix redirect logic in nginx.conf

Summary: The previous redirect implementation did not work as intended.

9muir reviewed (commented):
Looks pretty good, would be nice if there was an easy way to test it. Will confirm that it doesn't break our stuff.

9muir commented on cloud/configs/nginx.conf:
```diff
@@ -14,9 +14,13 @@ server {
     access_log /var/log/nginx/etagecom.io_access.log;
     error_log /var/log/nginx/etagecom.io_error.log;
 
+    #CANNOT redirect POST messages (they change to GETs). Reject instead.
```
nit: space after '#'

9muir commented on cloud/configs/nginx.conf:
```diff
@@ -14,9 +14,13 @@ server {
 
+    #CANNOT redirect POST messages (they change to GETs). Reject instead.
+    if ($http_x_forwarded_proto != "https" & $request_method = POST) {
+       return 405 Allow "GET" always;
```
'Allow "GET" always' seems like a cryptic response to me

9muir replied:
This line also causes an error:

`invalid number of arguments in "return" directive in /etc/nginx/sites-enabled/nginx.conf:19`

9muir reviewed (commented):
Tested in staging envs, no problems. Apart from those two nits this looks good.

9muir commented on cloud/configs/nginx.conf:
```diff
@@ -14,9 +14,13 @@ server {
     error_log /var/log/nginx/etagecom.io_error.log;
 
+    #CANNOT redirect POST messages (they change to GETs). Reject instead.
+    if ($http_x_forwarded_proto != "https" & $request_method = POST) {
```
I take my comment about 'looks good' back, the '&' operator seems to cause nginx problems, at least in my environment (nginx 1.4.6). Is this the syntax you tested?

9muir commented:
The problem of supporting multi-term conditionals in nginx seems relatively common, so there are probably various ways you could go to address the POST case. Or you could just split this into two separate PRs and we can merge the 'rewrite -> return 301' fix separately.

Here's a quick stab at this that seems to mostly work:

```    set $cond "";
    if ($http_x_forwarded_proto != "https") {
        set $cond "!https";
    }
    if ($request_method = "POST") {
        set $cond "${cond}:post";
    }
    if ($cond = "!https:post") {
       return 405 'Allow "GET" always';
    }
```

[Shamelessly taken from https://www.digitalocean.com/community/questions/conditional-within-an-if-in-nginx]

9muir commented on cloud/configs/nginx.conf:
```diff
@@ -14,9 +14,13 @@ server {
     # Force HTTPS
     if ($http_x_forwarded_proto != "https") {
-      rewrite ^(.*)$ https://$host$request_uri permanent;
+      return 301 https://$host$request_uri;
```
nit: should be indented 4 spaces

matt9j committed:
[cloud] Address spacing nits

The code style for this project prefers 4 spaces and spaces after "#"
in comments where possible per the pep-8 styleguide. Not mentioned in
review, but it seems like the file's style also uses a newline after
the closing brace of a conditional block.

matt9j committed:
Revert "[cloud] Reject non-https nginx posts"

This reverts commit a1b9ebb06b2a9eb77a41c2975db945d553357ade.

This hunk causes two errors in deployment and will have to be
re-worked. Removing for now to allow merging the required 301 fix in
the same pull request.

```
invalid number of arguments in "return" directive in /etc/nginx/sites-enabled/nginx.conf:19

invalid condition "$http_x_forwarded_proto" in /etc/nginx/sites-enabled/nginx.conf:18
```

9muir commented:
One other thing that I think will be useful is an exception for localhost, since we don't (at least in AWS) terminate SSL locally. That probably then rolls up into the general 'non-https' redirect conditional, something like this for the PR as a whole:

```    # Force HTTPS if not connecting from localhost
    set $cond "";
    if ($http_x_forwarded_proto != "https") {
        set $cond "!https";
    }
    if ($request_method = "POST") {
        set $cond "${cond}:post";
    }
    if ($remote_addr = "127.0.0.1") {
        set $cond "${cond}:localhost";
    }
    # the variable $cond encapsulates the 3 factors we care about in determining
    # how to handle non-https messages:
    # 1) accept only if they come from localhost - the suffix 'localhost' in the
    # $cond variable causes the two conditionals below to fail and be skipped.
    # 2) send standard 301 to any non-POST request (no 'post' suffix)
    # 3) send a 405 reponse to POST requests, since otherwise clients will turn
    # them into GET on receiving the 301 response.
    if ($cond = "!https") {
      return 301 https://$host$request_uri;
    }
    if ($cond = "!https:post") {
       return 405;
    }
```

Can you test against the UW env please?

Commit:
[cloud] Reject non-https posts from external clients

Give external clients an explicit 405 error when they attempt a
non-https post. Without this rejection the post will be redirected to
an https get and subsequently fail. This provides an explicit
exception for traffic originating from localhost since we do not
currently terminate ssl on localhost in the production system.

matt9j committed:
[cloud] Add a reload warning to nginx.conf

This commit is documentation only and should not change behavior.

matt-9-johnson commented:
> Can you test against the UW env please?

This passes a basic sniff test in the UW environment-- after reloading the configuration nginx runs fine and returns the appropriate values for the different conditions. I tested with curl that I got the expected 405 for an http remote post, 301 for an http remote get, and no server errors for localhost or https. I also did a quick sanity check of the site itself to make sure my instance was running stably.

matt-9-johnson commented on cloud/configs/nginx.conf:
```diff
@@ -5,6 +5,10 @@
 # of patent rights can be found in the PATENTS file in the same directory.
 
+# WARNING: Changes to this file will not be reflected until the
+# running nginx instance is explicity reloaded `nginx -s reload` or
```
Looking around I found out that reloading a config is a common webserver problem, and nginx actually supports an in-place config update signal that does not require a restart and does not interrupt any connections. If we think it's a good idea I can open another pull request to do this in-place reload with the fab deploy, but didn't want to hold this PR up with mixed changes.

9muir replied:
Thanks for looking into that Matt. I think having a way to non-disruptively restart nginx is very useful, in part because it will save other people from having to figure out how to do it manually. It could be either part of the postinst script, which is the approach I took and then abandoned, or it could be a step run by fab if the deploy is successful. I'd probably lean towards the former,  since then the config reload will take place if a deployment is done in some way other than using fab, e.g.., thru the AWS CodeDeploy console.
